drivers/platform/aducm3029/uart.c

```diff
@@ -0,0 +1,401 @@
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+/******************************************************************************/
+/************************* Include Files **************************************/
+/******************************************************************************/
+
+#include "uart.h"
+#include "uart_extra.h"
+#include <stdlib.h>
+
+/******************************************************************************/
+/*************************** Types Declarations *******************************/
+/******************************************************************************/
+
+/**
+ * @struct baud_desc
+ * @brief Structure of an element from \ref baud_rates_26MHz.
+ * The baud rate is calculated with the formula:
+ * Baudrate = (CLK_FREQ / (div_m + div_n/2048) * pow(2,osr+2) * div_c)).
+ */
+struct baud_desc {
+	/** Calculated baud rate from the following parameters */
+	uint32_t	baud_rate;
+	/** From 0 to 2047 */
+	uint16_t	div_n;
+	/** From 1 to 3 */
+	uint8_t		div_m;
+	/** From 1 to 65535 */
+	uint16_t	div_c;
+	/** From 0 to 3 */
+	uint8_t		osr;
+};
+
+
+/******************************************************************************/
+/********************** Macros and Constants Definitions **********************/
+/******************************************************************************/
+
+#define MAX_BYTES	1024u
+#define CLK_FREQ	26000000u
+#define BAUDS_NB	10u
+
+/** Stores calculated baud rates and parameters */
+const struct baud_desc baud_rates_26MHz[BAUDS_NB] = {
+	{9600, 1078, 3, 24, 3},
+	{19200, 1078, 3, 12, 3},
+	{38400, 1321, 3, 8, 2},
+	{57600, 1078, 3, 4, 3},
+	{115200, 1563, 1, 4, 3},
+	{230400, 1563, 1, 2, 3},
+	{460800, 1563, 1, 1, 3},
+	{921600, 1563, 1, 1, 2},
+	{1000000,1280, 1, 1, 2},
+	{1500000, 171, 1, 1, 2}
+};
+
+/**
+ * Used to check if device already initialized
+ */
+static uint32_t initialized;
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+/**
+ * @brief Allocates the memory needed for the UART descriptor
+ * @return Address to the allocated memory, NULL if the allocation fails.
+ */
+static struct uart_desc *alloc_desc_mem(void)
+{
+	struct uart_desc	*desc;
+	struct aducm_uart_desc	*aducm_desc;
+	uint32_t		mem;
+
+	desc = calloc(1, sizeof(struct uart_desc));
+	if (!desc)
+		return NULL;
+	aducm_desc = calloc(1, sizeof(struct aducm_uart_desc));
+	if (!aducm_desc) {
+		free(desc);
+		return NULL;
+	}
+	desc->extra = aducm_desc;
+
+	/* Allocate and align buffer to 32 bits */
+	aducm_desc->adi_uart_buffer = calloc(1, ADI_UART_BIDIR_MEMORY_SIZE + 3);
```

Comment:
What are the 3 extra bytes used for ?

Reply:
For example calloc return the address 0x1 and to align it to 32 bits (4 bytes) you need to add to the address 3 to make it 0x4.